### ISSUE TITLE
[Question] Limit learning rate

### ISSUE THREAD

**Issue by Specthram** [konstmish/prodigy#37]
Hello,

amazing work, thank you !

I've noticed that by increasing the number of epochs, the learning rate rises mechanically. Sometimes it can rise very high and I have the impression that it damages the trained lora.
Is there any way of limiting this learning rate? Currently, I use the d_coeff by setting it to 0.3 for example but I don't know if this is the best thing to do. maybe I'm missing something.
Thank you very much for your advices !

**Comment by qwe795138426** [comment]
Using cosine annealing, If you are training stable diffusion, do not use `scale weight norms`. Adjusting d_coef won't be that useful for your requirement

**Comment by qwe795138426** [comment]
from [https://github.com/LoganBooker/prodigy-plus-schedule-free](url), The original Prodigy's aim is not to act as a combined learning rate calculator and scheduler. It's meant to ballpark a good learning rate, and leave LR decay to your preferred scheduler (usually cosine).

**Comment by Specthram** [comment]
Thanks for your answer.

I forgot to mention, i'm using prodigy with kohya_ss gui, training stable diffusion xl based models, and i'm using cosine.
i'm also using scale weight norms set to 10 just to display the curve on tensorboard, to keep the weights values it around 0.7-1.0 but maybe I got a bad approach then.

**Comment by qwe795138426** [comment]
In my experience, scale weight norms will seriously affect the estimation of prodigy and make it overestimate. Usually, dropout and max_grad_norm, which are considered problematic, have little effect at low values, and noise-related settings will make it overestimate at the beginning. Don't get too hung up on tensorboard, try to cancel scale weight norms, maybe you will see the desired effect.

**Comment by qwe795138426** [comment]
[1740899006149.txt](https://github.com/user-attachments/files/19042023/1740899006149.txt)
These are the parameters I used for training, I hope it helps you(The eps in optimizer_args is a special setting for the variant)

**Comment by konstmish** [comment]
Hi @Specthram, thanks for sharing your experience using prodigy and your suggestion! It's actually quite easy to add the requested feature, we can definitely do it and maybe set float('inf') as the default value to avoid affecting the runs of others. I'm wondering what kind of values of the estimated learning rate do you consider too high?

**Comment by Specthram** [comment]
@qwe795138426 thank you for your for sharing your settings, I'll take a look at it :). Doesn't scale weight norm doesn't touch anything until it reaches the value (like 2) ? maybe I'm wrong. I stopped using it to see if there is a difference, it's hard to tell. What I know for sure, it's when the value is reach, and tries to go over it, "keys_scaled" is going up, and then yes the training seems to stop learning anything except artifacts and tends to destroys itself when going >= 1. In my case I set it very big (10) just to know how my lora strength is, 10 never be reached.

@konstmish hi ! I'm more of a beginner than someone with experience. Besides what I said in the first post, I can tell for sure that most of my issues are in my way of tagging and choose my pictures, or on the noise settings that I do not master properly.
For example, according to the complexity of the triggerword itself, for example "f5e1s2", it seems to learns better than a simpler word in some cases, like it merge the simpler word with something already known and quickly overtrain without learn the things I want.
In some other cases, something like "heart_shaped_rock" seems to learn more accuratly what I want to learn because it seems to understand that there is something about rock and shape. -but- It's very hard to tell the difference, and even more if there is something wrong in the training parameters or optimizer !

Edit : 
Ok I think I just totally missed your question :
this is a training where my values rises with my epoch count.
![Image](https://github.com/user-attachments/assets/edab3e6a-da70-46b2-8521-9858a2c8fb90)
are theses pictures giving you the informations you asked me ?

**Comment by konstmish** [comment]
Thanks for sharing the figures. If I understand correctly, the left figure shows the product of the estimated value of `d` times the learning rate coefficient from the cosine scheduler. The difference between the peaks seems to be roughly by a factor of 2. Combining these facts, I suspect that the estimated value of `d` is somewhat consistent between the training runs, and the difference in peaks is mostly due to the scheduler value at the time when `d` reaches its peak value (which seems to be at around 600-1000 iterations).

I suspect that capping the maximum value of the learning rate estimate would not make a significant difference in your case. It is possible that the degradation you see with longer training runs is due to overfitting (perhaps not enough data or the weight decay value is too small?) rather than due to learning rates being too large, though there might be other things at play here.

**Comment by Specthram** [comment]
Ok, I think I understand, I misread the graph. I thought, for example, at 1000 steps the lora is more or less trained depending of the target total steps, because of the higher curve.

**Comment by konstmish** [comment]
From your graphs, it looks like the learning rate reaches reasonable values ($10^{-5}$, which is a factor of ~3 smaller than the peak value) by iteration 200, but that doesn't mean anything about the quality of the lora weights. It depends on a lot of other factors such as how much data you have. For instance, when training on imagenet, which is a big dataset, we observe early convergence of the learning rate but we have to keep training for a very long time after that. On a small dataset, which is probably the case for you, it might be that you don't need that many iterations as you may want to avoid overfitting.

I'm closing this issue for now but feel free to reopen if you want to continue the discussion.
